### ISSUE TITLE
RKE2 Provisioning: Custom Clusters

### ISSUE THREAD

**Issue by deniseschannon** [rancher/rancher#32374]
Ability to create a custom RKE2 cluster

**Comment by thehejik** [comment]
The RKE2 custom provisioning doesn't work (yet?) in recent rancher:master-head `91a6f51`.

When executed as root the provided command which finish successfully (curl -k needed for untrusted certs):
```
curl -skfL https://rancher.hopto.org/system-agent-install.sh |  sh -s - --server https://rancher.hopto.org --token <REDACTED> --ca-checksum <REDACTED> --etcd --controlplane --worker
```
(Stdout [here](https://github.com/rancher/rancher/files/6426378/system-agent-install-stdout.log))


The systemd service `rancher-system-agent.service` is in failed state:
```
systemctl status rancher-system-agent.service
● rancher-system-agent.service - Rancher System Agent
   Loaded: loaded (/etc/systemd/system/rancher-system-agent.service; enabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Wed 2021-05-05 10:16:48 UTC; 2s ago
     Docs: https://www.rancher.com
  Process: 6838 ExecStart=/usr/bin/rancher-system-agent (code=exited, status=2)
 Main PID: 6838 (code=exited, status=2)

May 05 10:16:48 localhost systemd[1]: rancher-system-agent.service: Unit entered failed state.
May 05 10:16:48 localhost systemd[1]: rancher-system-agent.service: Failed with result 'exit-code'.
```

Logs from the failed unit:
```
# journalctl -u rancher-system-agent.service
-- Logs begin at Wed 2021-05-05 09:46:14 UTC, end at Wed 2021-05-05 10:20:08 UTC. --
May 05 10:05:31 localhost systemd[1]: Started Rancher System Agent.
May 05 10:05:31 localhost rancher-system-agent[3354]: time="2021-05-05T10:05:31Z" level=info msg="Rancher System Agent version v0.0.1-alpha15 (05>
May 05 10:05:31 localhost rancher-system-agent[3354]: time="2021-05-05T10:05:31Z" level=info msg="Using directory /var/lib/rancher/agent/work for>
May 05 10:05:31 localhost rancher-system-agent[3354]: time="2021-05-05T10:05:31Z" level=info msg="Starting remote watch of plans"
May 05 10:05:31 localhost rancher-system-agent[3354]: time="2021-05-05T10:05:31Z" level=info msg="Starting local watch of plans in /var/lib/ranch>
May 05 10:05:31 localhost rancher-system-agent[3354]: panic: invalid configuration: no server found for cluster "agent"
May 05 10:05:31 localhost rancher-system-agent[3354]: goroutine 9 [running]:
May 05 10:05:31 localhost rancher-system-agent[3354]: github.com/rancher/system-agent/pkg/remoteplan.(*watcher).start(0xc00040ca80, 0x1727060, 0x>
May 05 10:05:31 localhost rancher-system-agent[3354]:         /go/src/github.com/rancher/system-agent/pkg/remoteplan/watcher.go:44 +0x765
May 05 10:05:31 localhost rancher-system-agent[3354]: created by github.com/rancher/system-agent/pkg/remoteplan.Watch
May 05 10:05:31 localhost rancher-system-agent[3354]:         /go/src/github.com/rancher/system-agent/pkg/remoteplan/watcher.go:31 +0xb5
May 05 10:05:31 localhost systemd[1]: rancher-system-agent.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
May 05 10:05:31 localhost systemd[1]: rancher-system-agent.service: Unit entered failed state.
May 05 10:05:31 localhost systemd[1]: rancher-system-agent.service: Failed with result 'exit-code'.
```

It appears the problem is `panic: invalid configuration: no server found for cluster "agent"` message and following crash.

Further investigation is likely needed. 

Also the agent can be called from shell outside systemd with the same result as above:
```
# CATTLE_LOGLEVEL=debug CATTLE_AGENT_CONFIG=/etc/rancher/agent/config.yaml /usr/bin/rancher-system-agent
```

Systemd unit file (files used by `EnvironmentFile` keys are either empty or non-existing)
```
# cat /etc/systemd/system/rancher-system-agent.service
[Unit]
Description=Rancher System Agent
Documentation=https://www.rancher.com
Wants=network-online.target
After=network-online.target
[Install]
WantedBy=multi-user.target
[Service]
EnvironmentFile=-/etc/default/rancher-system-agent
EnvironmentFile=-/etc/sysconfig/rancher-system-agent
EnvironmentFile=-/etc/systemd/system/rancher-system-agent.env
Type=simple
Restart=always
RestartSec=5s
Environment=CATTLE_LOGLEVEL=debug
Environment=CATTLE_AGENT_CONFIG=/etc/rancher/agent/config.yaml
ExecStart=/usr/bin/rancher-system-agent
```

Content of /etc/rancher/agent/config.yaml:
```
# cat /etc/rancher/agent/config.yaml
workDirectory: /var/lib/rancher/agent/work
localPlanDirectory: /var/lib/rancher/agent/plans
appliedPlanDirectory: /var/lib/rancher/agent/applied
remoteEnabled: true
preserveWorkDirectory: false
connectionInfoFile: /var/lib/rancher/agent/rancher2_connection_info.json
```
Content of  /var/lib/rancher/agent/rancher2_connection_info.json:
```
# cat /var/lib/rancher/agent/rancher2_connection_info.json
{
  "kubeConfig": "apiVersion: v1\nclusters:\n- cluster:\n    certificate-authority-data: <REDACTED>\n    server: \"\"\n  name: agent\ncontexts:\n- context:\n    cluster: agent\n    user: agent\n  name: agent\ncurrent-context: agent\nkind: Config\npreferences: {}\nusers:\n- name: agent\n  user:\n    token: <REDACTED>\n",
  "namespace": "fleet-default",
  "secretName": "custom-7bf157929033-machine-plan"
}
```

**Comment by thehejik** [comment]
Problem is that server argument is not parsed correctly in `/var/lib/rancher/agent/rancher2_connection_info.json` file where the value is just empty string `server: \"\"\n`

**Comment by thehejik** [comment]
**Test report:**
Successfully validated custom RKE2 clusters provisioning using k8s `v1.21.1-rc3+rke2r1` from Vue (system resources checked, nothing else) on `rancher/rancher:master-head e756276`.

Following setup of nodes and its roles was tested:
- 1x cp+etcd+worker (all roles on one node)
- 3x cp+etcd, 3x worker
- 3x cp, 3x etcd, 3x worker

I was able delete these clusters from Vue after successful deployment - the clusters disappeared from Ember and Vue then.

During testing I've identified following issues:
* NEW https://github.com/rancher/dashboard/issues/3331 (Vue issue)
* NEW https://github.com/rancher/rancher/issues/33363 (I'd say RKE2 issue)

Closing.
